### DISCUSSION TITLE
How to distinguish MS assemblies from NuGet and bundled with the SDK

### DISCUSSION THREAD

**Discussion by Trivivium** [dotnet/runtime#50276] [General]
I'm working on a tool to assess the use of NuGet packages in a C# project. However, I have run into an issue regarding how to distinguish between packages published by Microsoft and those bundled with the SDK? In both cases, the namespace is either `System.*` or `Microsoft.*`, which means I can't use that reliably.

Is there maybe another way to extract this from the assembly metadata? The runtime itself? If the answer is there is no automatic way to determine this, do there exist an official list of packages bundled with the SDK?

The tool is not an analyzer or a source generator, but it is using Roslyn if that opens up some possibilities.

**Comment by huoyaoyuan** [comment]
![image](https://user-images.githubusercontent.com/5644458/112711793-04522480-8f06-11eb-96a2-8b54c23d6851.png)

Inspect this attribute? I'm not sure whether it's accurate.

**Reply by Trivivium** [reply]
I didn't know that attribute value existed, thanks!

Filtering by this attribute alone was enough to remove runtime assemblies reliably for my use-case. However, the attribute is not added to the `Microsoft.NET.Sdk.Web` SDK. But your answer led me to also discovering the `ReferenceAssemblyAttribute`. A combination of the two proved to be a sufficient filter.

**Update**

After further testing it seems the combination of `AssemblyMetadata(".NETFrameworkAssembly", "")` can result in false positives. An example is the: `Microsoft.Extensions.Logging.Abstractions`. As far as I can tell this is not bundled with the standard SDK.
